[ML] Add internal flag to flush api to control whether or not to flush indices

edsavage opened the pull request:
When datafeeds send flush requests they don't require the indices to be refreshed (and in fact may be detrimentally expensive in the case of small bucket sizes).

This change adds an (internal) flag to control whether or not a refresh is required when flushing.

edsavage committed:
[ML] Add internal flag to flush api determining whether to refresh indices or not.

When datafeeds send flush requests they don't require the indices to be refreshed.

First working version.

edsavage committed:
Tidy up

 * change naming
 * remove unnecessary logging

edsavage committed:
Merge branch 'main' of github.com:elasticsearch/elasticsearch into add_internal_flag_to_flush_api

edsavage committed:
Merge branch 'main' of github.com:elasticsearch/elasticsearch into add_internal_flag_to_flush_api

edsavage committed:
Merge branch 'main' of github.com:elasticsearch/elasticsearch into add_internal_flag_to_flush_api

edsavage committed:
AutodetectResultProcessorIT.testMultipleFlushesBetweenPersisting test fix

Always refresh indices in FlushAcknowlegement instance created within integration tests. This is to ensure consistent behaviour across test runs.

edsavage committed:
Merge branch 'main' of github.com:elasticsearch/elasticsearch into add_internal_flag_to_flush_api

edsavage committed:
Merge branch 'main' of github.com:elasticsearch/elasticsearch into add_internal_flag_to_flush_api

edsavage commented:
retest

edsavage committed:
Merge branch 'main' of github.com:elasticsearch/elasticsearch into add_internal_flag_to_flush_api

droberts195 commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/FlushJobAction.java:
```diff
@@ -88,6 +92,7 @@ public void writeTo(StreamOutput out) throws IOException {
             out.writeOptionalString(advanceTime);
             out.writeOptionalString(skipTime);
             out.writeBoolean(waitForNormalization);
+            out.writeBoolean(refreshRequired);
```
This needs BWC protection. Please read the JavaDocs in `TransportVersion.java` to see how to do it now.

droberts195 commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/FlushJobAction.java:
```diff
@@ -77,6 +80,7 @@ public Request(StreamInput in) throws IOException {
             advanceTime = in.readOptionalString();
             skipTime = in.readOptionalString();
             waitForNormalization = in.readBoolean();
+            refreshRequired = in.readBoolean();
```
Needs BWC protection.

droberts195 commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/process/autodetect/params/FlushJobParams.java:
```diff
@@ -93,6 +100,10 @@ public boolean isWaitForNormalization() {
         return waitForNormalization;
     }
 
+    public Boolean refreshRequired() {
```
```suggestion
    public boolean refreshRequired() {
```

edsavage committed:
Attend to review comments

 * use boolean in preference to Boolean
 * BWC guards

droberts195 commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/FlushJobAction.java:
```diff
@@ -187,6 +212,7 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
             if (skipTime != null) {
                 builder.field(SKIP_TIME.getPreferredName(), skipTime);
             }
+            builder.field(REFRESH_REQUIRED.getPreferredName(), refreshRequired);
```
Similarly, it shouldn't be included in the JSON representation. (Like `waitForNormalization` is not mentioned in this method.)

droberts195 commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/job/process/autodetect/output/FlushAcknowledgement.java:
```diff
@@ -28,44 +29,57 @@ public class FlushAcknowledgement implements ToXContentObject, Writeable {
     static {
         PARSER.declareString(ConstructingObjectParser.constructorArg(), ID);
         PARSER.declareLong(ConstructingObjectParser.optionalConstructorArg(), LAST_FINALIZED_BUCKET_END);
+        PARSER.declareBoolean(ConstructingObjectParser.optionalConstructorArg(), REFRESH_REQUIRED);
```
Here it _is_ correct to include the internal field in the parser, as this is for internal communications with our C++ code.

edsavage committed:
Attend to further review comments

 * Do not parse "refreshRequired" flag in FlushJobAction

edsavage committed:
Fix failing BWC tests

 * For backwards compatibility it must be possible to construct FlushAcknowledgement without refreshRequired, i.e. refreshRequired must be nullable.
 * Added missing comparisons in FlushParams equals and hashCode methods.

droberts195 reviewed (approved):
LGTM

edsavage committed:
Merge branch 'main' of github.com:elasticsearch/elasticsearch into add_internal_flag_to_flush_api

edsavage committed:
Fix failing BWC test

For transport versions prior to V_8_500_012 refreshRequired in FlushJobAction.Request is always set to true. Update the tests to reflect that.

edsavage committed:
Fix format violation
